File: MeteoPesca/WeatherService.swift
import Foundation

public struct FetchedWeatherData {
    public let waterTemp: Double
    public let cloudCover: Double
    public let windDirectionChange: Double
    public let swellHeight: Double
    public let surfaceTempDelta24h: Double
}

public class WeatherService {
    
    public static func fetchWeather(latitude: Double, longitude: Double) async throws -> FetchedWeatherData {
        // 1. Fetch Forecast Data (Atmospheric Conditions)
        let forecastUrlString = "https://api.open-meteo.com/v1/forecast?latitude=\(latitude)&longitude=\(longitude)&current=cloud_cover,wind_direction_10m&hourly=wind_direction_10m,temperature_2m&forecast_days=1"
        guard let forecastUrl = URL(string: forecastUrlString) else {
            throw URLError(.badURL)
        }
        
        let (forecastData, _) = try await URLSession.shared.data(from: forecastUrl)
        let forecastJSON = try JSONSerialization.jsonObject(with: forecastData) as? [String: Any]
        
        // Parse cloud cover
        let current = forecastJSON?["current"] as? [String: Any]
        let cloudCover = current?["cloud_cover"] as? Double ?? 20.0
        
        // Parse hourly values for wind direction change and air temperature
        let hourly = forecastJSON?["hourly"] as? [String: Any]
        let hourlyWind = hourly?["wind_direction_10m"] as? [Double] ?? []
        let hourlyAir = hourly?["temperature_2m"] as? [Double] ?? []
        
        // Wind direction change over last 3 hours (using index 12 vs index 9)
        var windDirectionChange = 10.0
        if hourlyWind.count > 12 {
            let change = abs(hourlyWind[12] - hourlyWind[9])
            windDirectionChange = change > 180.0 ? 360.0 - change : change
        }
        
        // Air temperature change over 24 hours (last index vs first index)
        var surfaceTempDelta24h = 0.0
        if hourlyAir.count >= 24 {
            surfaceTempDelta24h = hourlyAir[hourlyAir.count - 1] - hourlyAir[0]
        }
        
        // 2. Fetch Marine Data (Native Sea Surface Temperature & Swell Height)
        let marineUrlString = "https://marine-api.open-meteo.com/v1/marine?latitude=\(latitude)&longitude=\(longitude)&current=sea_surface_temperature,wave_height&forecast_days=1"
        var waterTemp = 20.0
        var swellHeight = 0.2
        
        if let marineUrl = URL(string: marineUrlString) {
            do {
                let (marineData, _) = try await URLSession.shared.data(from: marineUrl)
                if let marineJSON = try JSONSerialization.jsonObject(with: marineData) as? [String: Any],
                   let currentMarine = marineJSON["current"] as? [String: Any] {
                    if let sst = currentMarine["sea_surface_temperature"] as? Double {
                        waterTemp = sst
                    }
                    if let waveHeight = currentMarine["wave_height"] as? Double {
                        swellHeight = waveHeight
                    }
                }
            } catch {
                // If inland or marine API fails (e.g. coordinates inland), fallback to neutral defaults
                print("Marine API fetch failed or coordinates inland: \(error)")
            }
        }
        
        return FetchedWeatherData(
            waterTemp: waterTemp,
            cloudCover: cloudCover,
            windDirectionChange: windDirectionChange,
            swellHeight: swellHeight,
            surfaceTempDelta24h: surfaceTempDelta24h
        )
    }
}

File: previsioni-pesca-android/app/src/main/java/com/vincenzo/previsionipesca/engine/WeatherService.kt
package com.vincenzo.previsionipesca.engine

import com.vincenzo.previsionipesca.models.FetchedWeatherData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.abs

object WeatherService {

    private suspend fun getUrlContent(urlString: String): String = withContext(Dispatchers.IO) {
        val url = URL(urlString)
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "GET"
        conn.connectTimeout = 10000
        conn.readTimeout = 10000
        
        val code = conn.responseCode
        if (code != 200) {
            throw Exception("HTTP Error $code")
        }
        
        val reader = BufferedReader(InputStreamReader(conn.inputStream))
        val sb = StringBuilder()
        var line: String?
        while (reader.readLine().also { line = it } != null) {
            sb.append(line)
        }
        reader.close()
        conn.disconnect()
        sb.toString()
    }

    suspend fun fetch7DayWeather(latitude: Double, longitude: Double): Map<String, FetchedWeatherData> = withContext(Dispatchers.IO) {
        val forecastUrl = "https://api.open-meteo.com/v1/forecast?latitude=$latitude&longitude=$longitude&hourly=cloud_cover,wind_direction_10m,temperature_2m,wind_speed_10m&wind_speed_unit=ms&forecast_days=7"
        val marineUrl = "https://marine-api.open-meteo.com/v1/marine?latitude=$latitude&longitude=$longitude&hourly=sea_surface_temperature,wave_height&forecast_days=7"
        
        val forecastResponse = getUrlContent(forecastUrl)
        val forecastJson = JSONObject(forecastResponse)
        val hourlyForecast = forecastJson.optJSONObject("hourly")
        
        val clouds = optDoubleArray(hourlyForecast, "cloud_cover")
        val winds = optDoubleArray(hourlyForecast, "wind_direction_10m")
        val temps = optDoubleArray(hourlyForecast, "temperature_2m")
        val windSpeeds = optDoubleArray(hourlyForecast, "wind_speed_10m")
        
        // Parse time strings
        val timeStrings = mutableListOf<String>()
        val timeJsonArray = hourlyForecast?.optJSONArray("time")
        if (timeJsonArray != null) {
            for (i in 0 until timeJsonArray.length()) {
                timeStrings.add(timeJsonArray.optString(i, ""))
            }
        }

        var ssts = DoubleArray(0)
        var waves = DoubleArray(0)
        
        try {
            val marineResponse = getUrlContent(marineUrl)
            val marineJson = JSONObject(marineResponse)
            val hourlyMarine = marineJson.optJSONObject("hourly")
            ssts = optDoubleArray(hourlyMarine, "sea_surface_temperature")
            waves = optDoubleArray(hourlyMarine, "wave_height")
        } catch (e: Exception) {
            e.printStackTrace()
            // marine API can fail if coordinates are inland
        }
        
        val result = mutableMapOf<String, FetchedWeatherData>()
        
        for (day in 0 until 7) {
            val startIndex = day * 24
            val endIndex = (day + 1) * 24
            
            if (timeStrings.size <= startIndex) continue
            val dateStr = timeStrings[startIndex].take(10) // "yyyy-MM-dd"
            
            // Slice hourly values
            val cloudsSlice = sliceArray(clouds, startIndex, endIndex)
            val windsSlice = sliceArray(winds, startIndex, endIndex)
            val tempsSlice = sliceArray(temps, startIndex, endIndex)
            val sstSlice = sliceArray(ssts, startIndex, endIndex)
            val waveSlice = sliceArray(waves, startIndex, endIndex)
            val windSpeedSlice = sliceArray(windSpeeds, startIndex, endIndex)
            
            val avgCloud = if (cloudsSlice.isEmpty()) 20.0 else cloudsSlice.average()
            
            // Wind direction change over last 3 hours of mid-day
            var windChange = 10.0
            if (windsSlice.size > 12) {
                val change = abs(windsSlice[12] - windsSlice[9])
                windChange = if (change > 180.0) 360.0 - change else change
            }
            
            val tempDelta = if (tempsSlice.size >= 24) tempsSlice[23] - tempsSlice[0] else 0.0
            val avgSst = if (sstSlice.isEmpty()) 20.0 else sstSlice.average()
            val maxWave = if (waveSlice.isEmpty()) 0.2 else waveSlice.maxOrNull() ?: 0.2
            val avgWind = if (windSpeedSlice.isEmpty()) 4.0 else windSpeedSlice.average()
            
            result[dateStr] = FetchedWeatherData(
                waterTemp = avgSst,
                cloudCover = avgCloud,
                windDirectionChange = windChange,
                swellHeight = maxWave,
                surfaceTempDelta24h = tempDelta,
                windSpeedMps = avgWind
            )
        }
        
        result
    }

    private fun optDoubleArray(obj: JSONObject?, key: String): DoubleArray {
        val arr = obj?.optJSONArray(key) ?: return DoubleArray(0)
        val res = DoubleArray(arr.length())
        for (i in 0 until arr.length()) {
            res[i] = arr.optDouble(i, 0.0)
        }
        return res
    }

    private fun sliceArray(arr: DoubleArray, start: Int, end: Int): DoubleArray {
        if (arr.size <= start) return DoubleArray(0)
        val actualEnd = minOf(end, arr.size)
        return arr.copyOfRange(start, actualEnd)
    }
}
